operator/pkg/controllers/controller.go

```diff
@@ -59,7 +58,7 @@ func (c *GenericController) Reconcile(ctx context.Context, req reconcile.Request
 		resource.GetObjectKind().SetGroupVersionKind(v1alpha1.SchemeGroupVersion.WithKind(v1alpha1.ControlPlaneKind))
 	}
 	// 2. Copy object for merge patch base
-	persisted := resource.DeepCopyObject()
+	persisted := c.DeepCopy(resource)
```

Comment:
Why did you add this deep copy interface? It seems like you've wired it through a bunch of controllers, but are just casting. You might as well just cast here.

Reply:
In older versions for controller-runtime libraries `client.MergeFrom()` used to accept `runtime.Object` but in updated versions`client.MergeFrom()` accepts `client.Object`. `DeepCopyObject()` returns a `runtime.Object`.

Here we will have to cast both ControlPlane and Dataplane objects and if we at some point add more higher level objects to the controller this switch cases will grow.